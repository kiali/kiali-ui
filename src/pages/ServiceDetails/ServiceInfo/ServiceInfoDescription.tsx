import * as React from 'react';
import { Col, Row } from 'patternfly-react';
import LocalTime from '../../../components/Time/LocalTime';
import { DisplayMode, HealthIndicator } from '../../../components/Health/HealthIndicator';
import { ServiceHealth } from '../../../types/Health';
import { Endpoints } from '../../../types/ServiceInfo';
import { Port, ObjectValidation } from '../../../types/IstioObjects';
import { style } from 'typestyle';
import {
  ConfigIndicator,
  NOT_VALID,
  SMALL_SIZE,
  MEDIUM_SIZE
} from '../../../components/ConfigValidation/ConfigIndicator';
import { Popover, OverlayTrigger, Icon, Tooltip } from 'patternfly-react';
import './ServiceInfoDescription.css';
import Labels from '../../../components/Label/Labels';
import { ThreeScaleServiceRule } from '../../../types/ThreeScale';

interface ServiceInfoDescriptionProps {
  name: string;
  namespace: string;
  createdAt: string;
  resourceVersion: string;
  istioEnabled?: boolean;
  labels?: { [key: string]: string };
  selectors?: { [key: string]: string };
  type?: string;
  ip?: any;
  externalName?: string;
  ports?: Port[];
  endpoints?: Endpoints[];
  health?: ServiceHealth;
  threeScaleServiceRule?: ThreeScaleServiceRule;
  validations?: ObjectValidation;
}

const listStyle = style({
  listStyleType: 'none',
  padding: 0
});

const labelTitleStyle = style({
  marginBottom: '2px'
});

const labelListStyle = style({
  marginBottom: '4px'
});

const ExternalNameType = 'ExternalName';

class ServiceInfoDescription extends React.Component<ServiceInfoDescriptionProps> {
  getValidations(): ObjectValidation {
    return this.props.validations ? this.props.validations : ({} as ObjectValidation);
  }

  getPortOver(portId: number): Popover {
    return (
      <Popover id={portId + '-config-validation'} title={NOT_VALID.name} style={{ maxWidth: '80%', minWidth: '200px' }}>
        <div>{this.getPortIssue(portId)}</div>
      </Popover>
    );
  }

  getPortIssue(portId: number): string {
    let message = '';
    if (this.props.validations && this.props.validations.checks) {
      message = this.props.validations.checks
        .filter(c => c.path === 'spec/ports[' + portId + ']')
        .map(c => c.message)
        .join(',');
    }
    return message;
  }

  hasIssue(portId: number): boolean {
    return this.getPortIssue(portId) !== '';
  }

  render() {
    return (
      <div className="card-pf">
        <div className="card-pf-body">
          <Row>
            <Col xs={12} sm={6} md={5} lg={5}>
              <div id="labels">
                <div className={'progress-description ' + labelTitleStyle}>
                  <strong>Labels</strong>
                </div>
                <div className={'label-collection ' + labelListStyle}>
                  <Labels labels={this.props.labels || {}} />
                </div>
              </div>
              <div id="selectors">
                <div className={'progress-description ' + labelTitleStyle}>
                  <strong>Selectors</strong>
                </div>
                <div className={'label-collection ' + labelListStyle}>
                  <Labels labels={this.props.selectors || {}} />
                </div>
              </div>
              <div>
                <strong>Type</strong> {this.props.type ? this.props.type : ''}
              </div>
              {this.props.type !== ExternalNameType ? (
                <div>
                  <strong>IP</strong> {this.props.ip ? this.props.ip : ''}
                </div>
              ) : (
                <div>
                  <strong>ExternalName</strong> {this.props.externalName ? this.props.externalName : ''}
                </div>
              )}
              <div>
                <strong>Created at</strong> <LocalTime time={this.props.createdAt} />
              </div>
              <div>
                <strong>Resource Version</strong> {this.props.resourceVersion}
              </div>
              {this.props.threeScaleServiceRule && this.props.threeScaleServiceRule.threeScaleHandlerName !== '' && (
                <span>
                  Service linked with 3scale API Handler <i>{this.props.threeScaleServiceRule.threeScaleHandlerName}</i>
                </span>
              )}
            </Col>
            <Col xs={12} sm={4} md={2} lg={2}>
              <div className="progress-description">
                <ConfigIndicator
                  id={this.props.name + '-config-validation'}
                  validations={[this.getValidations()]}
                  size={MEDIUM_SIZE}
                />
                <strong>Ports</strong>
              </div>
              <ul className={listStyle}>
                {(this.props.ports || []).map((port, i) => (
                  <li key={'port_' + i}>
                    {this.hasIssue(i) ? (
                      <OverlayTrigger
                        placement={'right'}
                        overlay={this.getPortOver(i)}
                        trigger={['hover', 'focus']}
                        rootClose={false}
                      >
                        <span style={{ color: NOT_VALID.color }}>
                          <Icon
                            type="pf"
                            name="error-circle-o"
                            style={{ fontSize: SMALL_SIZE }}
                            className="health-icon"
                            tabIndex="0"
                          />
                        </span>
                      </OverlayTrigger>
                    ) : (
                      undefined
                    )}
                    {port.protocol} {port.name} ({port.port})
                  </li>
                ))}
              </ul>
            </Col>
            <Col xs={12} sm={6} md={2} lg={2}>
              <div className="progress-description">
                <strong>Endpoints</strong>
              </div>
              {(this.props.endpoints || []).map((endpoint, i) => (
                <Row key={'endpoint_' + i}>
                  <Col xs={12} sm={12} md={12} lg={12}>
                    <ul className={listStyle}>
                      {(endpoint.addresses || []).map((address, u) => {
                        const id = 'endpoint_' + i + '_address_' + u;
                        return (
                          <li key={id}>
                            <OverlayTrigger
                              overlay={<Tooltip id={id + '_tooltip'}>{address.name}</Tooltip>}
                              trigger={['hover', 'focus']}
                              rootClose={false}
                            >
                              <strong>{address.ip} </strong>
                            </OverlayTrigger>
                          </li>
                        );
                      })}
                    </ul>
                  </Col>
                </Row>
              ))}
            </Col>
            <Col xs={12} sm={6} md={3} lg={3}>
              <div className="progress-description">
                <strong>Health</strong>
              </div>
              <HealthIndicator
                id={this.props.name}
                health={this.props.health}
                mode={DisplayMode.LARGE}
                tooltipPlacement="left"
              />
            </Col>
          </Row>
        </div>
      </div>
    );
  }
}

export default ServiceInfoDescription;
