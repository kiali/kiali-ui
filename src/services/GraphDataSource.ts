import { AppenderString, DurationInSeconds, TimeInMilliseconds, TimeInSeconds } from '../types/Common';
import {
  DecoratedGraphElements,
  EdgeLabelMode,
  GraphDefinition,
  GraphElements,
  GraphType,
  GroupByType,
  NodeParamsType,
  NodeType,
  UNKNOWN,
  DecoratedGraphNodeWrapper
} from '../types/Graph';
import Namespace from '../types/Namespace';
import * as AlertUtils from '../utils/AlertUtils';
import { PromisesRegistry } from '../utils/CancelablePromises';
import * as API from './Api';
import { decorateGraphData } from '../store/Selectors/GraphData';
import EventEmitter from 'eventemitter3';
import { createSelector } from 'reselect';
import { NamespaceAppHealth, NamespaceServiceHealth, NamespaceWorkloadHealth, NA } from 'types/Health';

export const EMPTY_GRAPH_DATA = { nodes: [], edges: [] };
const PROMISE_KEY = 'CURRENT_REQUEST';

// GraphDataSource allows us to have multiple graphs in play, which functionally allows us to maintain
// the master graph page as well as to offer mini-graphs in the detail pages.
//
// GraphDataSource (GDS) emits events asynchronously and has the potential to disrupt the expected
// react+redux workflow typical of our components.  To avoid unexpected results here are some
// [anti-]patterns for using GraphDataSource:
//   - Do not set up GDS callbacks in nested components.  It is better to process the callbacks in the
//     top-level component and then update props (via react or redux) and let the lower components update normally.
//       - if A embeds B, do not have callbacks for the same GDS in A and B, just A
//   - Avoid accessing GDS fields to access fetch information (elements, timestamps, fetchParameters, etc).  In
//     short, the fields are volatile and can change at unexpected times.
//       - Instead, in the callbacks save what you need to local variables or properties.  Then use them to
//         trigger react/redux state changes normally.
//   - Avoid passing a GDS as a property.
//       - The only reason to do this is for an embedded component to access the GDS fields directly, which is
//         an anti-pattern explained above.  Having said that, if you are SURE the GDS is stable, it will work
//         (at this writing we still do this for mini-graphs).

type EmitEvents = {
  (eventName: 'loadStart', isPreviousDataInvalid: boolean, fetchParams: FetchParams): void;
  (eventName: 'emptyNamespaces', fetchParams: FetchParams): void;
  (eventName: 'fetchError', errorMessage: string | null, fetchParams: FetchParams): void;
  (
    eventName: 'fetchSuccess',
    graphTimestamp: TimeInSeconds,
    graphDuration: DurationInSeconds,
    graphData: DecoratedGraphElements,
    fetchParams: FetchParams
  ): void;
};

type NamespaceHealth = NamespaceAppHealth | NamespaceServiceHealth | NamespaceWorkloadHealth;
type NodeHealth = {
  key: string;
  node: DecoratedGraphNodeWrapper;
};

export interface FetchParams {
  duration: DurationInSeconds;
  edgeLabelMode: EdgeLabelMode;
  graphType: GraphType;
  includeHealth: boolean;
  injectServiceNodes: boolean;
  namespaces: Namespace[];
  node?: NodeParamsType;
  queryTime?: TimeInMilliseconds; // default now
  showIdleEdges: boolean;
  showIdleNodes: boolean;
  showOperationNodes: boolean;
  showSecurity: boolean;
}

type OnEvents = {
  (eventName: 'loadStart', callback: (isPreviousDataInvalid: boolean, fetchParams: FetchParams) => void): void;
  (eventName: 'emptyNamespaces', callback: (fetchParams: FetchParams) => void): void;
  (eventName: 'fetchError', callback: (errorMessage: string | null, fetchParams: FetchParams) => void): void;
  (
    eventName: 'fetchSuccess',
    callback: (
      graphTimestamp: TimeInSeconds,
      graphDuration: DurationInSeconds,
      graphData: DecoratedGraphElements,
      fetchParams: FetchParams
    ) => void
  ): void;
};

export default class GraphDataSource {
  public graphDuration: DurationInSeconds;
  public graphTimestamp: TimeInSeconds;

  private _errorMessage: string | null;
  private _fetchParams: FetchParams;
  private _isError: boolean;
  private _isLoading: boolean;

  private eventEmitter: EventEmitter;
  private graphElements: GraphElements;
  private promiseRegistry: PromisesRegistry;
  private decoratedData = createSelector(
    (graphData: GraphElements) => graphData,
    graphData => decorateGraphData(graphData)
  );

  // Public methods

  constructor() {
    this.graphElements = EMPTY_GRAPH_DATA;
    this.graphDuration = 0;
    this.graphTimestamp = 0;

    this.eventEmitter = new EventEmitter();
    this.promiseRegistry = new PromisesRegistry();

    this._errorMessage = null;
    this._fetchParams = {
      duration: 0,
      edgeLabelMode: EdgeLabelMode.NONE,
      graphType: GraphType.VERSIONED_APP,
      includeHealth: true,
      injectServiceNodes: true,
      namespaces: [],
      showIdleEdges: false,
      showIdleNodes: false,
      showOperationNodes: false,
      showSecurity: false
    };
    this._isError = this._isLoading = false;
  }

  public fetchGraphData = (fetchParams: FetchParams) => {
    const previousFetchParams = this.fetchParameters;

    // Copy fetch parameters to a local attribute
    this._fetchParams = { ...fetchParams };

    if (fetchParams.namespaces.length === 0) {
      this._isLoading = this._isError = false;
      this.graphElements = EMPTY_GRAPH_DATA;
      this.graphDuration = 0;
      this.graphTimestamp = 0;
      this.emit('emptyNamespaces', fetchParams);
      return;
    }

    const restParams: any = {
      duration: fetchParams.duration + 's',
      graphType: fetchParams.graphType,
      includeIdleEdges: fetchParams.showIdleEdges,
      injectServiceNodes: fetchParams.injectServiceNodes
    };

    if (fetchParams.graphType === GraphType.APP || fetchParams.graphType === GraphType.VERSIONED_APP) {
      restParams.groupBy = GroupByType.APP;
    }

    if (fetchParams.queryTime) {
      restParams.queryTime = String(Math.floor(fetchParams.queryTime / 1000));
    }

    // Some appenders are expensive so only specify an appender if needed.
    let appenders: AppenderString = 'deadNode,sidecarsCheck,serviceEntry,istio';

    if (fetchParams.showOperationNodes) {
      appenders += ',aggregateNode';
    }

    if (!fetchParams.node && fetchParams.showIdleNodes) {
      // note we only use the idleNode appender if this is NOT a drilled-in node graph and
      // the user specifically requests to see idle nodes.
      appenders += ',idleNode';
    }

    if (fetchParams.showSecurity) {
      appenders += ',securityPolicy';
    }

    switch (fetchParams.edgeLabelMode) {
      case EdgeLabelMode.RESPONSE_TIME_95TH_PERCENTILE:
        appenders += ',responseTime';
        break;

      case EdgeLabelMode.REQUEST_RATE:
      case EdgeLabelMode.REQUEST_DISTRIBUTION:
      case EdgeLabelMode.NONE:
      default:
        break;
    }
    restParams.appenders = appenders;

    this._isLoading = true;
    this._isError = false;

    const isPreviousDataInvalid =
      previousFetchParams.namespaces.map(ns => ns.name).join() !==
        this.fetchParameters.namespaces.map(ns => ns.name).join() ||
      previousFetchParams.node !== this.fetchParameters.node ||
      previousFetchParams.graphType !== this.fetchParameters.graphType ||
      previousFetchParams.includeHealth !== this.fetchParameters.includeHealth ||
      previousFetchParams.injectServiceNodes !== this.fetchParameters.injectServiceNodes ||
      previousFetchParams.showOperationNodes !== this.fetchParameters.showOperationNodes ||
      previousFetchParams.showIdleNodes !== this.fetchParameters.showIdleNodes;

    if (isPreviousDataInvalid) {
      // Reset the graph data
      this.graphElements = EMPTY_GRAPH_DATA;
      this.graphDuration = 0;
      this.graphTimestamp = 0;
    }

    this.emit('loadStart', isPreviousDataInvalid, fetchParams);
    if (fetchParams.node) {
      this.fetchDataForNode(restParams);
    } else {
      this.fetchDataForNamespaces(restParams);
    }
  };

  public on: OnEvents = (eventName: any, callback: any) => {
    this.eventEmitter.on(eventName, callback);
  };

  public removeListener: OnEvents = (eventName: any, callback: any) => {
    this.eventEmitter.removeListener(eventName, callback);
  };

  // Some helpers

  public fetchForApp = (duration: DurationInSeconds, namespace: string, app: string) => {
    const params = this.fetchForAppParams(duration, namespace, app);
    this.fetchGraphData(params);
  };

  public fetchForAppParams = (duration: DurationInSeconds, namespace: string, app: string): FetchParams => {
    const params = GraphDataSource.defaultFetchParams(duration, namespace);
    params.graphType = GraphType.APP;
    params.node!.nodeType = NodeType.APP;
    params.node!.app = app;
    return params;
  };

  public fetchForWorkload = (duration: DurationInSeconds, namespace: string, workload: string) => {
    const params = this.fetchForWorkloadParams(duration, namespace, workload);
    this.fetchGraphData(params);
  };

  public fetchForWorkloadParams = (duration: DurationInSeconds, namespace: string, workload: string): FetchParams => {
    const params = GraphDataSource.defaultFetchParams(duration, namespace);
    params.graphType = GraphType.WORKLOAD;
    params.node!.nodeType = NodeType.WORKLOAD;
    params.node!.workload = workload;
    return params;
  };

  public fetchForService = (duration: DurationInSeconds, namespace: string, service: string) => {
    const params = this.fetchForServiceParams(duration, namespace, service);
    this.fetchGraphData(params);
  };

  public fetchForServiceParams = (duration: DurationInSeconds, namespace: string, service: string): FetchParams => {
    const params = GraphDataSource.defaultFetchParams(duration, namespace);
    params.graphType = GraphType.WORKLOAD;
    params.node!.nodeType = NodeType.SERVICE;
    params.node!.service = service;
    return params;
  };

  public fetchForNamespace = (duration: DurationInSeconds, namespace: string) => {
    const params = this.fetchForNamespaceParams(duration, namespace);
    this.fetchGraphData(params);
  };

  public fetchForNamespaceParams = (duration: DurationInSeconds, namespace: string): FetchParams => {
    const params = GraphDataSource.defaultFetchParams(duration, namespace);
    params.graphType = GraphType.WORKLOAD;
    params.showSecurity = true;
    return params;
  };

  // Private methods

  private static defaultFetchParams(duration: DurationInSeconds, namespace: string): FetchParams {
    return {
      namespaces: [{ name: namespace }],
      duration: duration,
      graphType: GraphType.WORKLOAD,
      includeHealth: true,
      injectServiceNodes: true,
      edgeLabelMode: EdgeLabelMode.NONE,
      showIdleEdges: false,
      showIdleNodes: false,
      showOperationNodes: false,
      showSecurity: false,
      node: {
        app: '',
        namespace: { name: namespace },
        nodeType: NodeType.UNKNOWN,
        service: '',
        version: '',
        workload: ''
      }
    };
  }

  private emit: EmitEvents = (eventName: any, ...args) => {
    this.eventEmitter.emit(eventName, ...args);
  };

  private fetchDataForNamespaces = (restParams: any) => {
    restParams.namespaces = this.fetchParameters.namespaces.map(namespace => namespace.name).join(',');
    this.promiseRegistry.register(PROMISE_KEY, API.getGraphElements(restParams)).then(
      response => {
        const responseData: any = response.data;
        this.graphElements = responseData && responseData.elements ? responseData.elements : EMPTY_GRAPH_DATA;
        this.graphTimestamp = responseData && responseData.timestamp ? responseData.timestamp : 0;
        this.graphDuration = responseData && responseData.duration ? responseData.duration : 0;
        const decoratedGraphElements = this.graphData;
        if (this.fetchParameters.includeHealth) {
          this.fetchHealth(decoratedGraphElements);
        } else {
          this._isLoading = this._isError = false;
          this.emit(
            'fetchSuccess',
            this.graphTimestamp,
            this.graphDuration,
            decoratedGraphElements,
            this.fetchParameters
          );
        }
      },
      error => {
        this._isLoading = false;
        if (error.isCanceled) {
          return;
        }

        this._isError = true;
        this._errorMessage = API.getErrorString(error);
        AlertUtils.addError('Cannot load the graph', error);
        this.emit('fetchError', `Cannot load the graph: ${this.errorMessage}`, this.fetchParameters);
      }
    );
  };

  private fetchDataForNode = (restParams: any) => {
    this.promiseRegistry.register(PROMISE_KEY, API.getNodeGraphElements(this.fetchParameters.node!, restParams)).then(
      response => {
        const responseData: any = response.data;
        this.graphElements = responseData && responseData.elements ? responseData.elements : EMPTY_GRAPH_DATA;
        this.graphTimestamp = responseData && responseData.timestamp ? responseData.timestamp : 0;
        this.graphDuration = responseData && responseData.duration ? responseData.duration : 0;
        const decoratedGraphElements = this.graphData;
        if (this.fetchParameters.includeHealth) {
          this.fetchHealth(decoratedGraphElements);
        } else {
          this._isLoading = this._isError = false;
          this.emit(
            'fetchSuccess',
            this.graphTimestamp,
            this.graphDuration,
            decoratedGraphElements,
            this.fetchParameters
          );
        }
      },
      error => {
        this._isLoading = false;
        if (error.isCanceled) {
          return;
        }

        this._isError = true;
        this._errorMessage = API.getErrorString(error);
        AlertUtils.addError('Cannot load the graph', error);
        this.emit('fetchError', this.errorMessage, this.fetchParameters);
      }
    );
  };

  // Limit health fetches to only the necessary namespaces for the necessary types
  private fetchHealth = (decoratedGraphElements: DecoratedGraphElements) => {
    if (!decoratedGraphElements.nodes || decoratedGraphElements.nodes.length === 0) {
      this._isLoading = false;
      this.emit('fetchSuccess', this.graphTimestamp, this.graphDuration, decoratedGraphElements, this.fetchParameters);

      return;
    }

    const duration = this.fetchParameters.duration;
    const appNamespacePromises = new Map<string, Promise<NamespaceAppHealth>>();
    const serviceNamespacePromises = new Map<string, Promise<NamespaceServiceHealth>>();
    const workloadNamespacePromises = new Map<string, Promise<NamespaceWorkloadHealth>>();

    const promiseToNode = new Map<Promise<NamespaceHealth>, NodeHealth[]>();

    // Asynchronously fetch health
    for (const node of decoratedGraphElements.nodes) {
      // ignore nodes that can not have health calculated due lack of access or lack of info
      // note: UNKNOWN node is already marked inaccessible
      if (node.data.isInaccessible) {
        continue;
      }
      const namespace = node.data.namespace;
      const nodeType = node.data.nodeType;
      const workload = node.data.workload;
      const workloadOk = workload && workload !== '' && workload !== UNKNOWN;
      // use workload health when workload is set and valid (workload nodes or versionApp nodes)
      const useWorkloadHealth = nodeType === NodeType.WORKLOAD || (nodeType === NodeType.APP && workloadOk);

      if (useWorkloadHealth) {
        let promise = workloadNamespacePromises.get(namespace);
        const nodeHealth = { node: node, key: node.data.workload! };
        if (!promise) {
          promise = API.getNamespaceWorkloadHealth(namespace, duration);
          workloadNamespacePromises.set(namespace, promise);
          promiseToNode.set(promise, [nodeHealth]);
        } else {
          const nodeHealths = promiseToNode.get(promise);
          nodeHealths!.push(nodeHealth);
        }
      } else if (nodeType === NodeType.APP) {
        let promise = appNamespacePromises.get(namespace);
        const nodeHealth = { node: node, key: node.data.app! };
        if (!promise) {
          promise = API.getNamespaceAppHealth(namespace, duration);
          appNamespacePromises.set(namespace, promise);
          promiseToNode.set(promise, [nodeHealth]);
        } else {
          const nodeHealths = promiseToNode.get(promise);
          nodeHealths!.push(nodeHealth);
        }
      } else if (nodeType === NodeType.SERVICE) {
        let promise = serviceNamespacePromises.get(namespace);
        const nodeHealth = { node: node, key: node.data.service! };
        if (!promise) {
          promise = API.getNamespaceServiceHealth(namespace, duration);
          serviceNamespacePromises.set(namespace, promise);
          promiseToNode.set(promise, [nodeHealth]);
        } else {
          const nodeHealths = promiseToNode.get(promise);
          nodeHealths!.push(nodeHealth);
        }
      }
    }

    let healthPromises: Promise<NamespaceHealth>[] = Array.from(appNamespacePromises.values());
    healthPromises = healthPromises.concat(Array.from(serviceNamespacePromises.values()));
    healthPromises = healthPromises.concat(Array.from(workloadNamespacePromises.values()));

    new PromisesRegistry().registerAll('HEALTH_PROMISES', healthPromises).then(
      nsHealths => {
        nsHealths.forEach((nsHealth, i) => {
          promiseToNode.get(healthPromises[i])!.forEach(nh => {
            const health = nsHealth[nh.key];
            if (health) {
              nh.node.data.health = health;
              nh.node.data.healthStatus = health.getGlobalStatus().name;
            } else {
              nh.node.data.healthStatus = NA.name;
              console.debug(`No health found for [${nh.node.data.nodeType}] [${nh.key}]`);
            }
          });
        });

        this._isLoading = false;
        this.emit(
          'fetchSuccess',
          this.graphTimestamp,
          this.graphDuration,
          decoratedGraphElements,
          this.fetchParameters
        );
      },
      error => {
        this._isLoading = false;
        if (error.isCanceled) {
          return;
        }

        this._isError = true;
        this._errorMessage = API.getErrorString(error);
        AlertUtils.addError('Cannot load the graph [health]', error);
        this.emit('fetchError', this.errorMessage, this.fetchParameters);
      }
    );
  };

  // Getters and setters
  public get graphData(): DecoratedGraphElements {
    return this.decoratedData(this.graphElements);
  }

  public get graphDefinition(): GraphDefinition {
    return {
      duration: this.graphDuration,
      elements: this.graphElements,
      timestamp: this.graphTimestamp,
      graphType: this.fetchParameters.graphType
    };
  }

  public get errorMessage(): string | null {
    return this._errorMessage;
  }

  public get fetchParameters(): FetchParams {
    return this._fetchParams;
  }

  public get isError(): boolean {
    return this._isError;
  }

  public get isLoading(): boolean {
    return this._isLoading;
  }
}
