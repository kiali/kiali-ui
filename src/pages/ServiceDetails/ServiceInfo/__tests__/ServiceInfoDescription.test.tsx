import * as React from 'react';
import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import ServiceInfoDescription from '../ServiceInfoDescription';
import GraphDataSource from '../../../../services/GraphDataSource';

const labels = { app: 'reviews' };

const endpoints = [
  {
    addresses: [
      {
        kind: 'Pod',
        name: 'reviews-v2-4140793682-qrpm9',
        ip: '172.17.0.11'
      },
      {
        kind: 'Pod',
        name: 'reviews-v3-3651831602-zn9g6',
        ip: '172.17.0.14'
      },
      {
        kind: 'Pod',
        name: 'reviews-v1-401049526-tfstp',
        ip: '172.17.0.16'
      }
    ],
    ports: [
      {
        name: 'http',
        protocol: 'TCP',
        port: 9080
      }
    ]
  }
];

describe('#ServiceInfoDescription render correctly with data', () => {
  it('should render service description', () => {
    const miniGraphDS = new GraphDataSource();
    const wrapper = shallow(
      <ServiceInfoDescription
        name="reviews"
        namespace="my-namespace"
        labels={labels}
        type="ClusterIP"
        ip="172.30.78.33"
        endpoints={endpoints}
        createdAt="2018-04-04T15:11:46Z"
        resourceVersion="1234"
        additionalDetails={[]}
        miniGraphDatasource={miniGraphDS}
      />
    );
    expect(wrapper).toBeDefined();
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });
});
