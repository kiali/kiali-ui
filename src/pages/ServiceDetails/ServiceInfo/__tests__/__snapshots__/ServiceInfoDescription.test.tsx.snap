// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#ServiceInfoDescription render correctly with data should render service description 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ServiceInfoDescription
    createdAt="2018-04-04T15:11:46Z"
    endpoints={
      Array [
        Object {
          "addresses": Array [
            Object {
              "ip": "172.17.0.11",
              "kind": "Pod",
              "name": "reviews-v2-4140793682-qrpm9",
            },
            Object {
              "ip": "172.17.0.14",
              "kind": "Pod",
              "name": "reviews-v3-3651831602-zn9g6",
            },
            Object {
              "ip": "172.17.0.16",
              "kind": "Pod",
              "name": "reviews-v1-401049526-tfstp",
            },
          ],
          "ports": Array [
            Object {
              "name": "http",
              "port": 9080,
              "protocol": "TCP",
            },
          ],
        },
      ]
    }
    ip="172.30.78.33"
    labels={
      Object {
        "app": "reviews",
      }
    }
    name="reviews"
    resourceVersion="1234"
    type="ClusterIP"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "iconName": "service",
      "iconType": "pf",
      "istio": undefined,
      "items": <Row
        bsClass="row"
        componentClass="div"
      >
        <Col
          bsClass="col"
          componentClass="div"
          lg={2}
          md={2}
          sm={6}
          xs={12}
        >
          <div
            className="progress-description"
          >
            <strong>
              Labels
            </strong>
          </div>
          <div
            className="label-collection"
          >
            <div>
              <Label
                name="app"
                value="reviews"
              />
            </div>
          </div>
          <div>
            <strong>
              Type
            </strong>
             
            ClusterIP
          </div>
          <div>
            <strong>
              IP
            </strong>
             
            172.30.78.33
          </div>
          <div>
            <strong>
              Created at
            </strong>
             
            <LocalTime
              time="2018-04-04T15:11:46Z"
            />
          </div>
          <div>
            <strong>
              Resource Version
            </strong>
             
            1234
          </div>
        </Col>
        <Col
          bsClass="col"
          componentClass="div"
          lg={3}
          md={3}
          sm={6}
          xs={12}
        >
          <div
            className="progress-description"
          >
            <strong>
              Ports
            </strong>
          </div>
          <ul
            style={
              Object {
                "listStyleType": "none",
              }
            }
          />
        </Col>
        <Col
          bsClass="col"
          componentClass="div"
          lg={5}
          md={5}
          sm={6}
          xs={12}
        >
          <div
            className="progress-description"
          >
            <strong>
              Endpoints
            </strong>
          </div>
          <Row
            bsClass="row"
            componentClass="div"
          >
            <Col
              bsClass="col"
              componentClass="div"
              lg={12}
              md={12}
              sm={12}
              xs={12}
            >
              <ul
                style={
                  Object {
                    "listStyleType": "none",
                  }
                }
              >
                <li>
                  <strong>
                    172.17.0.11
                     
                  </strong>
                  : 
                  reviews-v2-4140793682-qrpm9
                </li>
                <li>
                  <strong>
                    172.17.0.14
                     
                  </strong>
                  : 
                  reviews-v3-3651831602-zn9g6
                </li>
                <li>
                  <strong>
                    172.17.0.16
                     
                  </strong>
                  : 
                  reviews-v1-401049526-tfstp
                </li>
              </ul>
            </Col>
          </Row>
        </Col>
        <Col
          bsClass="col"
          componentClass="div"
          lg={2}
          md={2}
          sm={6}
          xs={12}
        >
          <div
            className="progress-description"
          >
            <strong>
              Health
            </strong>
          </div>
          <HealthIndicator
            health={undefined}
            id="reviews"
            mode={0}
            tooltipPlacement="left"
          />
        </Col>
      </Row>,
      "title": "reviews",
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "iconName": "service",
        "iconType": "pf",
        "istio": undefined,
        "items": <Row
          bsClass="row"
          componentClass="div"
        >
          <Col
            bsClass="col"
            componentClass="div"
            lg={2}
            md={2}
            sm={6}
            xs={12}
          >
            <div
              className="progress-description"
            >
              <strong>
                Labels
              </strong>
            </div>
            <div
              className="label-collection"
            >
              <div>
                <Label
                  name="app"
                  value="reviews"
                />
              </div>
            </div>
            <div>
              <strong>
                Type
              </strong>
               
              ClusterIP
            </div>
            <div>
              <strong>
                IP
              </strong>
               
              172.30.78.33
            </div>
            <div>
              <strong>
                Created at
              </strong>
               
              <LocalTime
                time="2018-04-04T15:11:46Z"
              />
            </div>
            <div>
              <strong>
                Resource Version
              </strong>
               
              1234
            </div>
          </Col>
          <Col
            bsClass="col"
            componentClass="div"
            lg={3}
            md={3}
            sm={6}
            xs={12}
          >
            <div
              className="progress-description"
            >
              <strong>
                Ports
              </strong>
            </div>
            <ul
              style={
                Object {
                  "listStyleType": "none",
                }
              }
            />
          </Col>
          <Col
            bsClass="col"
            componentClass="div"
            lg={5}
            md={5}
            sm={6}
            xs={12}
          >
            <div
              className="progress-description"
            >
              <strong>
                Endpoints
              </strong>
            </div>
            <Row
              bsClass="row"
              componentClass="div"
            >
              <Col
                bsClass="col"
                componentClass="div"
                lg={12}
                md={12}
                sm={12}
                xs={12}
              >
                <ul
                  style={
                    Object {
                      "listStyleType": "none",
                    }
                  }
                >
                  <li>
                    <strong>
                      172.17.0.11
                       
                    </strong>
                    : 
                    reviews-v2-4140793682-qrpm9
                  </li>
                  <li>
                    <strong>
                      172.17.0.14
                       
                    </strong>
                    : 
                    reviews-v3-3651831602-zn9g6
                  </li>
                  <li>
                    <strong>
                      172.17.0.16
                       
                    </strong>
                    : 
                    reviews-v1-401049526-tfstp
                  </li>
                </ul>
              </Col>
            </Row>
          </Col>
          <Col
            bsClass="col"
            componentClass="div"
            lg={2}
            md={2}
            sm={6}
            xs={12}
          >
            <div
              className="progress-description"
            >
              <strong>
                Health
              </strong>
            </div>
            <HealthIndicator
              health={undefined}
              id="reviews"
              mode={0}
              tooltipPlacement="left"
            />
          </Col>
        </Row>,
        "title": "reviews",
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
